### DISCUSSION TITLE
how to set sqlite database url on windows

### DISCUSSION THREAD

**Discussion by dalekzhangdong** [strongswan/strongswan#875] [Q&A] [answered]
Hi, I run charon_svc.exe which configure with sqlite, when set sql plugin database URI in strongswan.conf, 
``` bash
    sql {
      database = sqlite://C:\\Users\\hellen\\Desktop\\swansqlite\\bin\\ipsec.db
    }
```
then run charon_svc，the log print:
sql plugin: database URI not set
so my question is how to set database url on windows platform?

**Comment by tobiasbrunner** [comment, marked as answer]
Did you put that snippet in the root section of strongswan.conf? Or correctly inside a `charon.plugins` section?

While the path might still be an issue (I actually don't know if such absolute paths will work), you shouldn't get that particular error because that's logged if the value is empty.

**Reply by dalekzhangdong** [reply]
yes, the strongswan.conf info is: 
``` bash
charon-svc {
    filelog { 
        txt {
            path = stdout
            flush_line = yes
        }
    }

    sql {
      database = sqlite://C:\\Users\\hellen\\Desktop\\swansqlite\\bin\\ipsec.db
    }
}
```
and configure with --enable-sql --enable-sqlite when build.
yestoday, I test alot,  in [sqlite documnet](https://www.sqlite.org/c3ref/open.html)  show about windows sqlit db file path should be like this:

![image](https://user-images.githubusercontent.com/24927522/152625189-7b1b9233-f57f-40f0-bc17-f76abb5d43fd.png)

it also failed log with sqlite url not set ,  I add some log in code for debug,

![image](https://user-images.githubusercontent.com/24927522/152626387-b19f5a85-e074-460d-9d0d-922323b33f43.png)

strongswan.conf database sqlite url whatever set, the url log print empty, and I set the window ipsec.db path to url in code directively,
it run ok, so I think it is a bug. 
in settings_lexer.l code file, the pattern about ":" 

![image](https://user-images.githubusercontent.com/24927522/152625554-e8401855-04a2-4539-829e-fae6825bd1a8.png)

because windows path contain ":" , may be this pattern deal not think about windows path case?

**Reply by tobiasbrunner** [reply]
> yes, the strongswan.conf info is:

Which is wrong, there is a `plugins` section missing around the `sql` section.


> because windows path contain ":" , may be this pattern deal not think about windows path case?

No, that rule doesn't apply when parsing values (after a `=` we are in state `val` so only the rules in `<val>` apply).

**Reply by dalekzhangdong** [reply]
Thanks for your help, you are right! 

``` bash 
charon-svc {
    filelog { 
        txt {
            path = stdout
            flush_line = yes
        }
    }
    start-scripts {
        swanctl-creds = swanctl --load-creds --noprompt
        swanctl-conns = swanctl --load-conns
    }
    plugins {
        sql {
          # /C:/Users/hellen/Desktop/swansqlite/bin/ipsec.db
          database = sqlite://./ipsec.db
        }
    }
}
```
I test with absolute and relative path, both way of url setting is fine.
